File: java/src/main/java/com/penapereira/patterns/memento/Originator.java
package com.penapereira.patterns.memento;

/** Creates mementos of its state and restores itself from them. */
public final class Originator {

    /** The snapshot. Only Originator can read the private field: the wide interface. */
    public static final class Memento {

        private final String state;

        private Memento(String state) {
            this.state = state;
        }
    }

    private String state = "";

    public void setState(String state) {
        this.state = state;
    }

    public String getState() {
        return state;
    }

    public Memento createMemento() {
        return new Memento(state);
    }

    public void restore(Memento memento) {
        state = memento.state;
    }
}

File: javascript/src/memento/originator.js
import { Memento } from "./memento.js";

/** Creates mementos of its state and restores itself from them. */
export class Originator {
  #state = "";

  setState(state) {
    this.#state = state;
  }

  getState() {
    return this.#state;
  }

  createMemento() {
    return new Memento(this.#state);
  }

  restore(memento) {
    this.#state = memento.state;
  }
}
